### PR TITLE
Include package variant with optimization for x86-64-v3

### PR DESCRIPTION
**Pull request by jjhelmus**
Include a variant on linux-64 and win-64 that has optimizations for the x86-64-v3 feature level. This variant includes optimization that use the AVX and AVX2 instructions set. This variant is restricted from being installed on incompatible system via a dependency on the `_x86_64-microarch-level >=3` package. 

Same as #7 but not from a fork and enables optimizations in CUDA variants as well.

### CODE REVIEW

**Comment by jjhelmus** [comment]
I'm guessing that the Prefect builds are failing because the [`_x86_64-microarch-level`](https://github.com/conda-forge/microarch-level-feedstock) packages are needed.

**Comment by cbouss** [comment]
Adding x86_64-microarch-level https://github.com/AnacondaRecipes/microarch-level-feedstock/pull/1

I am pretty sure it will fail though as we don't have archspec installed in the base environment of the builders.

**Review by cbouss — APPROVED** [review]

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
This looks good to me based on the current knowledge I have. I did not look deeply at the build number + string.

@jjhelmus @cbouss if you want this to make it to anaconda.org, you'll probably want to either merge in a different branch than main or use `channels` in abs.yaml instead of `upload_channels`.

**Comment by danpetry** [comment]
@jjhelmus if you're familiar with the microarch feedstock already , and have good enough confidence in the variants etc, would you mind reviewing the mentioned feedstock?

File: recipe/build.sh
Comment (danpetry) [inline recipe/build.sh:48]: do we need to change the LLAMA_AVX flags below as well?

File: recipe/meta.yaml
Comment (danpetry) [inline recipe/meta.yaml:30]: gpu_variant won't be metal for x86_64, that's an arm-only variant. I guess it should be
```
(gpu_variant == "metal") or ((gpu_variant == "cuda-11") and (x86_64_opt == "v3"))
```
?

Comment (danpetry) [reply recipe/meta.yaml:30]: and same for the line below

File: recipe/meta.yaml
Comment (danpetry) [inline recipe/meta.yaml:70]: this will then install the v3 variant on x86_64_v4 processors. the binaries will work fine, but it could lead to confusion for the future? could s/v3/v34 in the recipe and the build string to be clearer, for example

**Review by danpetry — COMMENTED** [review]
If it's on ai-staging, is it internal use only? ai navigator is public now, so I've reviewed this with that in mind.

File: recipe/build.sh
Comment (jjhelmus) [reply recipe/build.sh:48]: You can but they do not need to be set. When the LLAMA_* flags are set, cmake includes the associated compiler flags (`-mavx`, etc) when calling the compiler. The change to `-march` accomplishes the same results since `x86-64-v3` enables the same instruction sets and macros as the instruction set specific compiler flags.

Comment (cbouss) [reply recipe/build.sh:48]: Setting march=x86-64-v3 is sufficient from my understanding.

File: recipe/meta.yaml
Comment (cbouss) [reply recipe/meta.yaml:70]: The build number magic is taking care of that. In a way the no opt variant could be v01234.

File: recipe/meta.yaml
Comment (cbouss) [reply recipe/meta.yaml]: I updated the selectors. The end result should be the same though.

**Review by JeanChristopheMorinPerso — APPROVED** [review]

File: recipe/build.sh
Comment (danpetry) [reply recipe/build.sh:48]: ok, it's a bit confusing for future people not familiar with it so would prefer it to be removed but that's nonfunctional

**Comment by cbouss** [comment]
Holding off merging until we get a green light. I'd like a confirmation on how the package is consumed before pushing to the channel.

File: recipe/meta.yaml
Comment (danpetry) [reply recipe/meta.yaml:70]: well, the v3 variant will install preferentially on v4 cpus. So the opt=none variant is more like v012?

**Review by danpetry — APPROVED** [review]

File: recipe/meta.yaml
Comment (cbouss) [reply recipe/meta.yaml:70]: opt=none would be installable anywhere really

File: recipe/meta.yaml
Comment (intentionally-left-nil) [inline recipe/meta.yaml:37]: To make the targeting for this easier (e.g. in an environment.yml file) can you make the string `cpu_v0_`
Otherwise `cpu_*` will match both cpu_v3_ and cpu_

Comment (jjhelmus) [reply recipe/meta.yaml:37]: I would vote for `cpu_v1_` :smile:

Comment (cbouss) [reply recipe/meta.yaml]: added v1

**Pull request merged**
